### PR TITLE
feat: Update the data export process to include extra data files (M2-8485)

### PR DESCRIPTION
**Pull request by sultanofcardio**
- [x] Tests for the changes have been added
- [ ] OSS packages added to MindLogger [open source credit page](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/263127186/Admin+Panel+Applet+Builder+Library)

### 📝 Description

🔗 [Jira Ticket M2-8485](https://mindlogger.atlassian.net/browse/M2-8485)

This PR introduces two new file exports:
- Schedule history
- Flow item history

These new files should be present everywhere exporting data is currently available in the admin app. Including:
- Dashboard respondents tab (`/dashboard/respondents`) using the applet context menu
- Applet page top-level export button (e.g. on the overview page at `/dashboard/{applet_id}/overview`)
- Applet page activities tab (`/dashboard/{applet_id}/activities`) using the activity/flow context menu
- Applet page participants tab (`/dashboard/{applet_id}/participants`) using the participant table row context menu
- Participant detail page assignments tab (`/dashboard/{applet_id}/participants/{subject_id}`) using the activity/flow/participant context menu
- DataViz Summary tab (`/dashboard/{applet_id}/participants/{subject_id}/activities/{activity_id}/summary`) export button
- Applet Settings (`/dashboard/{applet_id}/settings`) export data button

> [!NOTE]
> The files will only be exported when the `enable-ema-extra-files` feature flag is set to `true`

#### Flow Item History

Given a date range, a file called `flow_activity_history.csv` is produced containing the history of each version of each activity within each version of each activity flow inside a given applet, even if they have since been deleted.

Columns:
- `applet_id`: Applet ID
- `flow_item_history_created_date`: Date that the flow item was added to the flow
- `activity_flow_id`: Flow ID
- `activity_flow_name`: Flow Name
- `applet_version`: Applet version number
- `activity_id`: Activity ID
- `activity_name`: Activity Name

> [!NOTE]
> See the [ticket](https://mindlogger.atlassian.net/browse/M2-8485) description for a sample `flow_activity_history.csv` file

#### Schedule History

Given a date range, a new file called `schedule_history.csv` is produced containing an enumeration of the history of scheduled events (both default and individual) between the start and end of the date range (inclusive) for every user in a given applet. An enumeration here means that for every day in the range (inclusive), we use the periodicity of each event version to project its availability on that given day for every user in the applet. This is a look back in time to approximate which days in the period each event version was available to the user.

For example, given the date range 2025-03-09 to 2025-03-15, we will look back in time at the following days to determine which events apply to these days. Days with no events will not show up in the file
- 2025-03-09
- 2025-03-10
- 2025-03-11
- 2025-03-12
- 2025-03-13
- 2025-03-14
- 2025-03-15

Using this example, if an ALWAYS available event was created on 2025-03-13,  there should be rows in the file for:
- 2025-03-13
- 2025-03-14
- 2025-03-15. 

However, a WEEKDAYS available event created on the same day (2025-03-13) will only have rows for
- 2025-03-13
- 2025-03-14

Columns:
- `applet_id`: Applet ID
- `applet_version`: Applet version number
- `user_id`: User ID
- `target_or_source_id`: The subject ID of the user identified by user_id
- `schedule_id`: Event ID
- `schedule_type`: activity or flow
- `schedule_version`: Event version number in the format YYYYMMDD-n, where n is the nth change on a given date
- `schedule_version_created_timestamp`: The time when this version of the event was created in UTC timezone
- `schedule_updated_by`: The user ID of the logged-in admin user who updated the event and created this version
- `activity_or_flow_id`: Activity ID or Flow ID based on what’s in event_type
- `activity_or_flow_name`: Activity name or Flow name based on what’s in event_type
- `activity_or_flow_hidden`: Whether the activity or flow is hidden in this version of the applet (values yes or no)
- `schedule_start_date`: Start date of the event periodicity
- `schedule_end_date`: End date of the event periodicity
- `schedule_date`: The given date within the period
- `schedule_start_time`: Event start time on a given date within the period
- `schedule_end_time`: Event end time on a given date within the period
- `access_before_schedule_start_time`: Whether the admin has enabled “Access before from time” on the event (values yes or no)
- `mobile_device_schedule_version`: The latest version of this event that this user has on any of their mobile devices
- `mobile_device_schedule_start_date`: Start date of the event periodicity from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_end_date`: End date of the event periodicity from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_start_time`: The start time from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_end_time`: The end time from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_access_before_start_time`: The value of the “Access before from time” flag on the version of the event on the user’s mobile device (values yes or no)
- `mobile_device_schedule_download_timestamp`: The date and time when the user downloaded this version of the event in UTC timezone
- `mobile_device_schedule_utc_timezone_offset`: The number of minutes to be added to the mobile_device_schedule_download_timestamp to convert it to the user’s timezone (for comparison with event start and end times)

> [!NOTE]
> See the [ticket](https://mindlogger.atlassian.net/browse/M2-8485) description for a sample `schedule_history.csv` file

This is an incredibly complicated endeavour with a lot of scenarios and edge cases. I have tried to flesh some of these out in a test file that hasn't been committed yet. However, the following scenarios have been identified as some important ones for the exported files:

- the schedule changes on the server **before** the previously scheduled event, and the device gets it immediately
- the schedule changes on the server **after** the previously scheduled event, and the device gets it immediately
- the schedule changes on the server **during** the previously scheduled event, and the device gets it immediately
- the schedule changes on the server **before** the previously scheduled event, and the device gets it **after the end, but still on the same day** of the previously scheduled event
- the schedule changes on the server **before** the previously scheduled event, and the device gets it **during** of the previously scheduled event
- the schedule changes on the server **before** the previously scheduled event, and the device gets it **after the end of the previously scheduled event...in fact, several days later**

> [!NOTE]
> See [MindLogger Schedule History Export](https://docs.google.com/spreadsheets/d/1lHWiwN1__e0j6REMW6ylY4wTxeG9We4UOlBINePr8W0/edit?gid=898155703#gid=898155703) Google Sheets file for samples of these scenarios

### 📸 Screenshots

N/A

### 🪤 Peer Testing

N/A

### ✏️ Notes

1. I have decided to use the `luxon` library for date/time calculations as part of this PR because I'm more familiar with it and it was already being included in the project as a transitive dependency. Usually I would have ultimately converted these calculations back to our current library `date-fns`, but I have decided not to do that in the interest of time.

### CODE REVIEW

File: src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx [hunk 1]
```diff
@@ -135,6 +139,26 @@ export const DataExportPopup = ({
...
+          if (ownerId) {
+            await new ScheduleHistoryExporter(ownerId).exportData({
+              appletId,
+              fromDate,
+              toDate,
+            });
+          } else {
+            console.warn(
+              `No owner ID found for current workspace. Schedule history export skipped.`,
+            );
```
Comment (mbanting) [inline src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx:176]: In what cases does this occur, out of curiosity?

File: src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+};
+
+type FullAccountParticipant = Omit<ParticipantWithDataAccess, 'id'> & { id: string };
+
+type ScheduleHistoryExportOptions = DataExporterOptions & {
+  subjectIds?: string;
+  respondentIds?: string;
+};
+
+export class ScheduleHistoryExporter extends DataExporter<
```
Comment (mbanting) [inline src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts:72]: Unlike `FlowItemHistoryExporter.ts` this is more complex and not as obvious what this is exporting. Recommend a brief description similar to what you provided in this PR description so that it's captured as part of the code, something like "...exports an enumeration of the history of scheduled events (both default and individual) between the start and end of the date range (inclusive) for every user in a given applet. "

File: src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+    const sortedDeviceScheduleHistoryData = unsortedDeviceScheduleHistoryData.sort(
+      (a, b) => new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
+    );
+
+    const periodStartDate = new Date(
+      params?.fromDate ?? sortedDeviceScheduleHistoryData[0].createdAt,
+    );
+    // const periodEndDate = new Date(params?.toDate ?? DateTime.now().toISO());
+    const periodEndDate = DateTime.fromISO('2025-03-14').toJSDate();
```
Comment (mbanting) [inline src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts]: Leftover mock data?

File: src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        );
+
+        schedulesForDay.forEach((schedule) => {
+          const deviceSchedule = this.findLatestMobileSchedule(
+            participant.id,
+            schedule,
+            sortedDeviceScheduleHistoryData,
+          );
+
+          // TODO: Figure out which schedule applies here, individual vs default
```
Comment (mbanting) [inline src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts]: Is this still to be done?

**Review by mbanting — CHANGES_REQUESTED** [review]
Took a first pass at the code. A challenging task nicely broken down and organized. Left a couple minor of comments, and I know the tests are still being written. Thank you as well for the extensive PR description. Super helpful. Great job!

File: src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx (hunk 1 above)
Comment (sultanofcardio) [reply src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx:176]: I guess if the redux workspace data fails to be populated

image

So technically it should never happen, since if it did then the applet wouldn't load to get to this point. This is just me handling the type error

File: src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts (hunk 2 above)
Comment (sultanofcardio) [reply src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts:72]: Good call. I still have a bit of cleanup to do. Thanks for the suggestion 🙏

File: src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts (hunk 3 above)
Comment (sultanofcardio) [reply src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts]: Indeed. Thanks for noticing this

File: src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts (hunk 4 above)
Comment (sultanofcardio) [reply src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts]: At the time, yes. I've since addressed this as one of the outstanding items while writing the test cases

File: src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx (hunk 1 above)
Comment (mbanting) [reply src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx:176]: Understood, thanks!

**Comment by sultanofcardio** [comment]
I've updated this PR to do a few additional things:
- Indicate whether an activity/flow is hidden in the export file (this one depends on https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1787)
- Account for events that extend past the end of the day (e.g. start at 10pm today and end at 2am tomorrow)
- Group events by applet version to avoid comparison between otherwise identical events

**Review by mbanting — APPROVED** [review]
Focused my re-review mostly on `ScheduleHistoryExporter`, it's tests, and the updated `daysBetweenInterval` function.

Very comprehensive implementation driven and validated by the impressive number of scenarios and their tests. Impressive!
